coreneuron/io/phase1.cpp

```diff
@@ -0,0 +1,170 @@
+#include <cassert>
+
+#include "coreneuron/nrniv/nrniv_decl.h"
+#include "coreneuron/sim/multicore.hpp"
+#include "coreneuron/io/phase1.hpp"
+#include "coreneuron/utils/nrnoc_aux.hpp"
+
+int (*nrn2core_get_dat1_)(int tid,
+                          int& n_presyn,
+                          int& n_netcon,
+                          int*& output_gid,
+                          int*& netcon_srcgid);
+
+namespace coreneuron {
+// no gid in any file can be greater than maxgid. maxgid will be set so that
+// maxgid * nrn_setup_multiple < 0x7fffffff
+static int maxgid = 0x7fffffff / nrn_setup_multiple;
+
+void Phase1::read_direct(int thread_id) {
+    int* output_gids;
+    int* netcon_srcgid;
+    int n_presyn;
+    int n_netcon;
+    int valid =
+        (*nrn2core_get_dat1_)(thread_id, n_presyn, n_netcon, output_gids, netcon_srcgid);
+    if (!valid) {
+        return;
+    }
+
+    this->output_gids = std::vector<int>(output_gids, output_gids + n_presyn);
+    delete[] output_gids;
+    this->netcon_srcgids = std::vector<int>(netcon_srcgid, netcon_srcgid + n_netcon);
+    delete[] netcon_srcgid;
```

Comment:
This is more a question to @nrnhines : In direct memory transfer mode between NEURON and CoreNEURON, do we rely on some memory addresses being maintained after first copy on coreneuron side?

For example:
* lets say we copied mechanism data to coreneuron through direct memory mode
* on coreneuron side we changed the data to new memory location and so on
* when coreneuron simulation finishes and neuron tries to copy back the data (e.g. voltages) or ml->data, are there any expectations in terms of where data is?

I think not but just want to double check.

Reply:
Above reminds me that we should add a test to copy some mechanism property (e.g. hh.gna) in https://github.com/BlueBrain/CoreNeuron/blob/76026e0b9820d96524ea3dc0e5c1d9f738e1f55a/tests/jenkins/neuron_direct.py#L22.

@nrnhines : In the above python script, how to access / copy mechanism's RANGE and GLOBAL from CoreNEURON side? cc: @iomaganaris